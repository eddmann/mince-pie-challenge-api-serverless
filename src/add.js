// @flow

import add from './handlers/add';
import { addPie } from 'db';
import createUserTokenAuthenticator from 'userTokenAuthenticator';

const { TABLE_NAME, USER_POOL_ID } = process.env;

if (!TABLE_NAME) {
  throw new Error('TABLE_NAME is not present');
}

if (!USER_POOL_ID) {
  throw new Error('USER_POOL_ID is not present');
}

export const handler = add({
  getUserIdFromToken: createUserTokenAuthenticator(USER_POOL_ID),
  addPie: addPie(TABLE_NAME),
});
