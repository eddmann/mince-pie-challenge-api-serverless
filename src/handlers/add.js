/* @flow */

import uuid from 'uuid';
import { badRequest, created, json } from '../helpers/http';
import { createHandler, strictHttpAuthentication } from '../helpers/handlers';
import { Resource } from 'hal';

const add = async ({ event, callback, userId, services: { db } }) => {
  const { name } = json(event.body);

  if (!name) {
    callback(
      undefined,
      badRequest('Invalid request body', [{ name: 'name', reason: "You must supply the pie's name." }]),
    );
    return;
  }

  const pie = {
    Id: uuid.v1(),
    UserId: userId,
    Name: name,
    AvgRating: 0,
    PhotoUrl: undefined,
    ThumbnailUrl: undefined,
    Ratings: {},
    AddedAt: new Date().getTime(),
  };

  await db.add(pie);

  const resource = new Resource(
    {
      id: pie.Id,
      name: pie.Name,
      rating: { avg: 0, total: 0 },
      addedAt: new Date(pie.AddedAt).toISOString(),
    },
    `/pies/${pie.Id}`,
  );

  resource.link('photo', `/pies/${pie.Id}/photo`);

  callback(undefined, created(resource));
};

export default createHandler(strictHttpAuthentication(add));
