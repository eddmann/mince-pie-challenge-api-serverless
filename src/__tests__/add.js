import add from '../handlers/add';

const response = () =>
  function cb(error, response) {
    if (error) fail(`Unexpected error: ${error}`);
    cb.response = response;
  };

it('requires an authenticated user', async () => {
  const services = { getUserIdFromToken: () => Promise.resolve() };
  const callback = response();

  await add(services)({ headers: { Authorization: '' } }, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});

it('requires a pie name', async () => {
  const services = { getUserIdFromToken: () => Promise.resolve('USER_ID') };
  const callback = response();

  await add(services)({ headers: { Authorization: 'TOKEN' } }, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});

it('adds a new mince pie', async () => {
  const services = {
    getUserIdFromToken: () => Promise.resolve('USER_ID'),
    db: { add: jest.fn() },
  };
  const callback = response();
  const body = JSON.stringify({ name: 'Mince Pie' });

  await add(services)({ headers: { Authorization: 'TOKEN' }, body }, {}, callback, undefined);

  expect(callback.response.statusCode).toEqual(201);
  expect(services.db.add).toBeCalled();
  const newPie = services.db.add.mock.calls[0][0];
  expect(newPie.UserId).toEqual('USER_ID');
  expect(newPie.Name).toEqual('Mince Pie');
});
