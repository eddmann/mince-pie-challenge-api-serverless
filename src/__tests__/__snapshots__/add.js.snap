// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requires a pie name 1`] = `
Object {
  "body": "{\\"title\\":\\"Bad Request\\",\\"detail\\":\\"Invalid request body\\",\\"errors\\":[{\\"name\\":\\"name\\",\\"reason\\":\\"You must supply the pie's name.\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/problem+json",
  },
  "statusCode": 400,
}
`;

exports[`requires an authenticated user 1`] = `
Object {
  "body": "{\\"title\\":\\"Forbidden\\",\\"detail\\":\\"Service requires authenticated user\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/problem+json",
  },
  "statusCode": 403,
}
`;
